core/src/common.cpp

```diff
@@ -207,6 +218,119 @@ namespace irods::http
+  }
+  
+    auto hit_intro_endpoint(std::string _encoded_body) -> nlohmann::json
+	{
+		const auto introspection_endpoint{
+			irods::http::globals::oidc_endpoint_configuration().at("introspection_endpoint").get_ref<const std::string&>()};
+
+		const auto parsed_uri{boost::urls::parse_uri(introspection_endpoint)};
+
+		if (parsed_uri.has_error()) {
+			log::error(
+				"{}: Error trying to parse introspection_endpoint [{}]. Please check configuration.", __func__, introspection_endpoint);
+			return {{"error", "bad endpoint"}};
+		}
+
+		const auto url{*parsed_uri};
+		const auto port{get_port_from_url(url)};
+
+		// Addr
+		net::io_context io_ctx;
+		auto tcp_stream{irods::http::transport_factory(url.scheme_id(), io_ctx)};
+		tcp_stream->connect(url.host(), *port);
+
+		const auto format_bearer_token{[](std::string_view _client_id, std::string_view _client_secret){
+		  auto encode_me{fmt::format("{}:{}", encode(_client_id), encode(_client_secret))};
+		  return safe_base64_encode(encode_me);
+		}};
+
+		auto client_id{irods::http::globals::oidc_configuration().at("client_id").get_ref<const std::string&>()};
+		auto client_secret{irods::http::globals::oidc_configuration().at("client_secret").get_ref<const std::string&>()};
+		auto auth_string{fmt::format("Basic {}", format_bearer_token(client_id, client_secret))};
+	
+		// Build Request
+		constexpr auto version_number{11};
+		beast::http::request<beast::http::string_body> req{beast::http::verb::post, url.path(), version_number};
+		req.set(beast::http::field::host, url.host());
+		req.set(beast::http::field::user_agent, irods::http::version::server_name);
+		req.set(beast::http::field::content_type,
+		        "application/x-www-form-urlencoded"); // Possibly set a diff way?
+		req.set(beast::http::field::authorization, auth_string);
+		req.set(beast::http::field::accept, "application/json");
+
+		req.body() = std::move(_encoded_body);
+		req.prepare_payload();
+
+		// Send request & recieve response
+		auto res{tcp_stream->communicate(req)};
+		
+		log::debug("{}: Got the following resp back: [{}]", __func__, res.body());
+
+		// JSONize response
+		return nlohmann::json::parse(res.body());
+	}
+
+	auto is_uuid(std::string_view _token) -> bool {
+		boost::uuids::string_generator gen{};
+		try {
+			auto possible_uuid{gen(_token.data())}; // This throws
+			return possible_uuid.version() != boost::uuids::uuid::version_type::version_unknown;
+		} catch (...) {
+			return false;
+		}
+	}
+
+	namespace json {
+		template <class InputIt, class UnaryPredicate>
+		constexpr auto find_if_not(InputIt first, InputIt last, UnaryPredicate p) -> InputIt {
+			while (first != last) {
+				if (p(first)) {
+					return first;
+				}
+				++first;
+			}
+			return last;
+		}
+
+		template <class InputIt, class UnaryPredicate>
+		constexpr auto all_of(InputIt first, InputIt last, UnaryPredicate p) -> bool {
+			return irods::http::json::find_if_not(first, last, p) == last;
+		}
+	} // namespace json
```

Comment:
Artifacts of me wanting `all_of` to work with json iterators.

Really, only line 320 is what needed to be changed. We don't want a receive a `*first` as our argument.

Could always not do the functional way too.

Reply:
What is the problem you're trying to solve with this?

Reply:
When you dereference a json iterator, you don't get the 'key' value, which is needed for mapping.

By not dereferencing the iterator in `find_if_not` on line 320, we get the information we need (a key-value pair).

Link to portion of code where I use the funtions defined:
https://github.com/irods/irods_client_http_api/blob/66e8f61475d0eed61bca5598a89475a91f1cd7db/core/src/common.cpp#L379-L396

Reply:
Turns out I was holding the json library wrong. Fixed.